src/vm/comutilnative.cpp

```diff
@@ -2087,60 +2088,65 @@ static INT32 RegularGetValueTypeHashCode(MethodTable *mt, void *pObjRef)
         //
         // <TODO> check this approximation - we may be losing exact type information </TODO>
         ApproxFieldDescIterator fdIterator(mt, ApproxFieldDescIterator::INSTANCE_FIELDS);
-        INT32 count = (INT32)fdIterator.Count();
 
-        if (count != 0)
+        FieldDesc *field;
+        while ((field = fdIterator.Next()) != NULL)
         {
-            for (INT32 i = 0; i < count; i++)
+            _ASSERTE(!field->IsRVA());
+            if (field->IsObjRef())
             {
-                FieldDesc *field = fdIterator.Next();
-                _ASSERTE(!field->IsRVA());
-                void *pFieldValue = (BYTE *)pObj + field->GetOffsetUnsafe();
-                if (field->IsObjRef())
+                // if we get an object reference we get the hash code out of that
+                if (*(Object**)((BYTE *)pObjRef + field->GetOffsetUnsafe()) != NULL)
```

Comment:
I wonder what was the reason for removing the pFieldValue and using the `(BYTE *)pObjRef + field->GetOffsetUnsafe()` at all places.

Reply:
I would have to add more GC protects for the pFieldValue. The original code tried to step around the GC protect, but it is not easy to do in the extra code that I have added (and I am also not sure whether the original got it 100% right).